# GRADED ASSIGNMENT FOR WEEK3 Segmenting and Clustering Neighborhoods in Toronto

<H2>PART ONE - CREATE PANDAS DATAFRAME WITH POSTAL CODES, BOROUGHS, AND NEIGHBOURHOODS

<b>Import some code from "Segmenting and Clustering Neighborhoods in New York City" lab

In [5]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0   conda-forge
    geopy:         1.18.1-py_0 conda-forge

geographiclib- 100% |################################| Time: 0:00:00  21.15 MB/s
geopy-1.18.1-p 100% |################################| Time: 0:00:00  30.83 MB/s
Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    altair:  2.2.2-py35_1 conda-forge
    branca:  0.3.1-py_0   conda-forge
    folium:  0.5.0-py_0   conda-forge
    vincent: 0.4.4-py_1   conda-forge

altair-2.2.2-p 100% |################################| Time: 0:00:00  54.39 MB/s
branca-0.3.1-p 100% |################################| Time: 0:00:00  24.40 MB/s
vincent-0.4.4- 100% |###################

<b>Import BeautfulSoup to scrape Wikipedia page that lists Toronto neighborhoods and postal codes

In [136]:
from bs4 import BeautifulSoup as BS
import urllib.request

<b>Now turn Wikipedia table into a dataframe

In [137]:
res = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
soup = BS(res.content,'lxml')
table = soup.find_all('table')[0] 
df = pd.read_html(str(table))[0]
postcode = df[0].tolist()
borough = df[1].tolist()
neigh = df[2].tolist()

df.head(10)

0                 1                 2
0  Postcode           Borough     Neighbourhood
1       M1A      Not assigned      Not assigned
2       M2A      Not assigned      Not assigned
3       M3A        North York         Parkwoods
4       M4A        North York  Victoria Village
5       M5A  Downtown Toronto      Harbourfront
6       M5A  Downtown Toronto       Regent Park
7       M6A        North York  Lawrence Heights
8       M6A        North York    Lawrence Manor
9       M7A      Queen's Park      Not assigned

<b>Delete false header row 0 - Can rename headers later

In [138]:
df = df.drop(0)

<b>Now eliminate rows with unassigned postcode

In [139]:
df = df.drop(df[df[1] == "Not assigned"].index)

<b>Now replace "Not assigned" Neighbourhood values with Borough value

In [140]:
df[2] = df[2].replace("Not assigned", df[1])

<b>Now combine neighbourhoods in the same postcode/borough into one record with Neighbourhood values comma delimited

In [141]:
df = pd.DataFrame({0:df[0],1:df[1],2:df[2]})

df = df.groupby([0,1])[2].apply(lambda tags : ', '.join(tags)).reset_index()

<b>Now rename columns as "PostalCode" "Borough" and "Neighbourhood"

In [142]:
df = df.rename(columns = {0:"PostalCode", 1:"Borough", 2:"Neighbourhood"})

df

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
38         M4G         East York   
39         M4H         East York   
40         M4J         East York   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
62         M5M        North York   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                         Neighbourhood  
0                                       Rouge, Malvern  
1               Highland Creek, Rouge Hill, Port Union  
2                    Guildwood, Morningside, West Hill  
3                              

<b>Use .shape method to display number of rows in dataframe

In [143]:
df.shape

(103, 3)

<H3>Dataframe has 103 rows and 3 columns

<H2>PART TWO - ADD LATITUDE AND LONGITUDE VALUES TO DATAFRAME

<b>Read .csv table of Toronto postal code lat/lon into dataframe

In [144]:
dfzip = pd.read_csv('https://cocl.us/Geospatial_data')

<b>Join latitude and longitude dataframe with postal code dataframe

In [145]:
dfp = df.join(dfzip)

<b>Drop redundant 'Postal Code' column

In [146]:
dfp = dfp.drop('Postal Code', axis=1)

dfp

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
38         M4G         East York   
39         M4H         East York   
40         M4J         East York   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
62         M5M        North York   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                       Rouge, Malvern  43.806686 -79.194353  
1               Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                    G

<b>Confirm that dataframe is correctly changed with .shape method

In [147]:
dfp.shape

(103, 5)

<h3>Dataframe is still 103 rows and now has 5 columns, latitude and longitude have been added

# PART THREE - EXPLORE AND CLUSTER NEIGHBOURHOODS

<h3>From here I "replicate the same analysis we did to the New York City data" as per
<br>the assignment instructions under "My Submission"

<h2>NOTE: I have converted the code imported from the ungraded assignment to understand it

<b>NOTE: I worked through each step scrutinizing all of the code to understand how it relates to the work I had already done

In [148]:
dftor = dfp[dfp['Borough'].str.contains("Toronto")]

dftor = dftor.drop(['PostalCode'], axis=1).reset_index(drop=True)

dftor

Borough                                      Neighbourhood  \
0       East Toronto                                        The Beaches   
1       East Toronto                       The Danforth West, Riverdale   
2       East Toronto                     The Beaches West, India Bazaar   
3       East Toronto                                    Studio District   
4    Central Toronto                                      Lawrence Park   
5    Central Toronto                                   Davisville North   
6    Central Toronto                                 North Toronto West   
7    Central Toronto                                         Davisville   
8    Central Toronto                        Moore Park, Summerhill East   
9    Central Toronto  Deer Park, Forest Hill SE, Rathnelly, South Hi...   
10  Downtown Toronto                                           Rosedale   
11  Downtown Toronto                        Cabbagetown, St. James Town   
12  Downtown Toronto                               Church and Wellesley   
13  Downtown Toronto                          Harbourfront, Regent Park   
14  Downtown Toronto                           Ryerson, Garden District   
15  Downtown Toronto                                     St. James Town   
16  Downtown Toronto                                        Berczy Park   
17  Downtown Toronto                                 Central Bay Street   
18  Downtown Toronto                           Adelaide, King, Richmond   
19  Downtown Toronto  Harbourfront East, Toronto Islands, Union Station   
20  Downtown Toronto           Design Exchange, Toronto Dominion Centre   
21  Downtown Toronto                     Commerce Court, Victoria Hotel   
22   Central Toronto                                           Roselawn   
23   Central Toronto                Forest Hill North, Forest Hill West   
24   Central Toronto                The Annex, North Midtown, Yorkville   
25  Downtown Toronto                     Harbord, University of Toronto   
26  Downtown Toronto          Chinatown, Grange Park, Kensington Market   
27  Downtown Toronto  CN Tower, Bathurst Quay, Island airport, Harbo...   
28  Downtown Toronto                    Stn A PO Boxes 25 The Esplanade   
29  Downtown Toronto             First Canadian Place, Underground city   
30  Downtown Toronto                                           Christie   
31      West Toronto                       Dovercourt Village, Dufferin   
32      West Toronto                           Little Portugal, Trinity   
33      West Toronto       Brockton, Exhibition Place, Parkdale Village   
34      West Toronto                      High Park, The Junction South   
35      West Toronto                             Parkdale, Roncesvalles   
36      West Toronto                                 Runnymede, Swansea   
37      East Toronto  Business Reply Mail Processing Centre 969 Eastern   

     Latitude  Longitude  
0   43.676357 -79.293031  
1   43.679557 -79.352188  
2   43.668999 -79.315572  
3   43.659526 -79.340923  
4   43.728020 -79.388790  
5   43.712751 -79.390197  
6   43.715383 -79.405678  
7   43.704324 -79.388790  
8   43.689574 -79.383160  
9   43.686412 -79.400049  
10  43.679563 -79.377529  
11  43.667967 -79.367675  
12  43.665860 -79.383160  
13  43.654260 -79.360636  
14  43.657162 -79.378937  
15  43.651494 -79.375418  
16  43.644771 -79.373306  
17  43.657952 -79.387383  
18  43.650571 -79.384568  
19  43.640816 -79.381752  
20  43.647177 -79.381576  
21  43.648198 -79.379817  
22  43.711695 -79.416936  
23  43.696948 -79.411307  
24  43.672710 -79.405678  
25  43.662696 -79.400049  
26  43.653206 -79.400049  
27  43.628947 -79.394420  
28  43.646435 -79.374846  
29  43.648429 -79.382280  
30  43.669542 -79.422564  
31  43.669005 -79.442259  
32  43.647927 -79.419750  
33  43.636847 -79.428191  
34  43.661608 -79.464763  
35  43.648960 -79.456325  
36  43.651571 -79.484450  
37  43.662744 -79.321558

In [149]:
dftor.shape

(38, 4)

<B>Obtain lat/long coordinates for Toronto

In [150]:
address = 'Toronto'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

/opt/conda/envs/DSX-Python35/lib/python3.5/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


The geograpical coordinates of Toronto are 43.653963, -79.387207.


<b>Produce a Folium map plotting Neighbourhood in Boroughs that contain "Toronto"

In [185]:
map_tor = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map

for i in range(0,len(dftor)):
    folium.Marker([dftor.iloc[i]['Latitude'], dftor.iloc[i]['Longitude']], popup=dftor.iloc[i]['Neighbourhood']).add_to(map_tor)

map_tor

<b>Make connecting to FourSquare easier

In [152]:
CLIENT_ID = 'ZGU4DXB3ROK4GNWTUAMJCARJMMC3J120HKWIYMXM52KJPZED' # your Foursquare ID
CLIENT_SECRET = 'RXGSRIHMW1EE1E54XL4NRHFIPJELLQ3SLFFZGW4UEDJ0VZTU' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ZGU4DXB3ROK4GNWTUAMJCARJMMC3J120HKWIYMXM52KJPZED
CLIENT_SECRET:RXGSRIHMW1EE1E54XL4NRHFIPJELLQ3SLFFZGW4UEDJ0VZTU


<b>EXPLORING A NEIGHBORHOOD
<br>We can look at Neighbourhood by random number between 0 & 102

In [153]:
from random import randint

neighnum = randint(0,37)
                
dftor.loc[neighnum, 'Neighbourhood']

'Church and Wellesley'

In [154]:
neigh_lat = dftor.loc[neighnum, 'Latitude'] # neighborhood latitude value
neigh_long = dftor.loc[neighnum, 'Longitude'] # neighborhood longitude value

neigh_name = dftor.loc[neighnum, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neigh_name, 
                                                               neigh_lat, 
                                                               neigh_long))

Latitude and longitude values of Church and Wellesley are 43.6658599, -79.38315990000001.


<b>Now, let's get up to 100 venues that are within a radius of 1000 meters.

In [155]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 1000 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neigh_lat, 
    neigh_long, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=ZGU4DXB3ROK4GNWTUAMJCARJMMC3J120HKWIYMXM52KJPZED&client_secret=RXGSRIHMW1EE1E54XL4NRHFIPJELLQ3SLFFZGW4UEDJ0VZTU&v=20180605&ll=43.6658599,-79.38315990000001&radius=1000&limit=100'

In [156]:
results = requests.get(url).json()
#results

In [157]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [158]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

name                     categories  \
0                        DanceLifeX Centre                   Dance Studio   
1                                    Smith                 Breakfast Spot   
2                       Sansotei Ramen 三草亭               Ramen Restaurant   
3                                The Alley                Bubble Tea Shop   
4                                DAVIDsTEA                       Tea Room   
5                        Glad Day Bookshop                      Bookstore   
6                    Ho's Team Barber Shop             Salon / Barbershop   
7                         Storm Crow Manor                            Bar   
8                           Wickson Social                      Gastropub   
9                        Barbara Hall Park                           Park   
10              North of Brooklyn Pizzeria                    Pizza Place   
11                             Tokyo Grill            Japanese Restaurant   
12                                Fabarnak                     Restaurant   
13                               Statler's                      Piano Bar   
14                                   FUEL+                      Juice Bar   
15                               401 Games                     Hobby Shop   
16                            Como En Casa             Mexican Restaurant   
17                    Churchmouse & Firkin                            Pub   
18                               Starbucks                    Coffee Shop   
19                           T-Swirl Crepe                       Creperie   
20                  Coach House Restaurant                          Diner   
21                      Woody's and Sailor                        Gay Bar   
22                  Hero Certified Burgers                   Burger Joint   
23                         Ethiopian House           Ethiopian Restaurant   
24  The Keg Steakhouse + Bar - Keg Mansion                     Steakhouse   
25                             Black Eagle                        Gay Bar   
26                             Riddle Room          General Entertainment   
27                               Seduction                 Adult Boutique   
28                    Toronto Hemp Company                     Smoke Shop   
29                   Kothur Indian Cuisine              Indian Restaurant   
30                          The Men's Room                    Men's Store   
31                             Tokyo Sushi               Sushi Restaurant   
32                  Curry's Art Store Ltd.            Arts & Crafts Store   
33                                    Wish                     Restaurant   
34                       Out On The Street                    Men's Store   
35                            Fresh Burger                   Burger Joint   
36                                 Loblaws                    Supermarket   
37                          Baskin Robbins                 Ice Cream Shop   
38                         Hair Of The Dog                      Gastropub   
39                      The Yoga Sanctuary                    Yoga Studio   
40                                  Ginger          Vietnamese Restaurant   
41                        Burrito Bandidos                  Burrito Place   
42                           Coffee Island                    Coffee Shop   
43                  Church Street Espresso                           Café   
44                                    YMCA                            Gym   
45                              Holy Chuck                   Burger Joint   
46                            Naan & Kabob              Afghan Restaurant   
47                 Cawthra Square Dog Park                        Dog Run   
48         Canadian National Ballet School                   Dance Studio   
49                                 Kintaro                       Sake Bar   
50                             Sabai Sabai                Thai Restaurant   
51                              Wo

In [159]:
print('{} venues returned by Foursquare.'.format(nearby_venues.shape[0]))

100 venues returned by Foursquare.


In [160]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [161]:
toronto_venues = getNearbyVenues(names=dftor['Neighbourhood'],
                                   latitudes=dftor['Latitude'],
                                   longitudes=dftor['Longitude']
                                  )


The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The 

In [162]:
print(toronto_venues.shape)
toronto_venues.head()

(1712, 7)


Neighbourhood  Neighbourhood Latitude  \
0                   The Beaches               43.676357   
1                   The Beaches               43.676357   
2                   The Beaches               43.676357   
3                   The Beaches               43.676357   
4  The Danforth West, Riverdale               43.679557   

   Neighbourhood Longitude                  Venue  Venue Latitude  \
0               -79.293031    Grover Pub and Grub       43.679181   
1               -79.293031              Starbucks       43.678798   
2               -79.293031  St-Denis Studios Inc.       43.675031   
3               -79.293031          Upper Beaches       43.680563   
4               -79.352188               Pantheon       43.677621   

   Venue Longitude    Venue Category  
0       -79.297215               Pub  
1       -79.298045       Coffee Shop  
2       -79.288022       Music Venue  
3       -79.292869      Neighborhood  
4       -79.351434  Greek Restaurant

In [163]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide, King, Richmond                                               100   
Berczy Park                                                             55   
Brockton, Exhibition Place, Parkdale Village                            18   
Business Reply Mail Processing Centre 969 Eastern                       17   
CN Tower, Bathurst Quay, Island airport, Harbou...                      14   
Cabbagetown, St. James Town                                             49   
Central Bay Street                                                      81   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                16   
Church and Wellesley                                                    89   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              37   
Davisville North                                                         9   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                      14   
Design Exchange, Toronto Dominion Centre                               100   
Dovercourt Village, Dufferin                                            20   
First Canadian Place, Underground city                                 100   
Forest Hill North, Forest Hill West                                      4   
Harbord, University of Toronto                                          35   
Harbourfront East, Toronto Islands, Union Station                      100   
Harbourfront, Regent Park                                               48   
High Park, The Junction South                                           24   
Lawrence Park                                                            4   
Little Portugal, Trinity                                                66   
Moore Park, Summerhill East                                              3   
North Toronto West                                                      21   
Parkdale, Roncesvalles                                                  16   
Rosedale                                                                 4   
Roselawn                                                                 2   
Runnymede, Swansea                                                      38   
Ryerson, Garden District                                               100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         95   
Studio District                                                         40   
The Annex, North Midtown, Yorkville                                     26   
The Beaches                                                              4   
The Beaches West, India Bazaar                                          21   
The Danforth West, Riverdale                                            42   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Adelaide, King, Richmond                                                100   
Berczy Park                                                              55   
Brockton, Exhibition Place, Parkdale Village                             18   
Business Reply Mail Processing Centre 969 Eastern                        17   
CN Tower, Bathurst Quay, Island airport, Harbou...                       14   
Cabbagetown, St. James Town                                              49   
Central Bay Street                                                       81   
Chinatown, Grange Park, Kensington Market                               100   
Christie                                                                 16   
Church and Wellesley                                          

In [164]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 236 uniques categories.


In [165]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Adult Boutique  Afghan Restaurant  Airport  \
0                   The Beaches               0                  0        0   
1                   The Beaches               0                  0        0   
2                   The Beaches               0                  0        0   
3                   The Beaches               0                  0        0   
4  The Danforth West, Riverdale               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                 0                    0             0         0   
1                 0                    0             0         0   
2                 0                    0             0         0   
3                 0                    0             0         0   
4                 0                    0             0         0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                 0            0           0                    0   
1                 0            0           0                    0   
2                 0            0           0                    0   
3                 0            0           0                    0   
4                 0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       0     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Beer Bar  Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  \
0         0           0                   0       0              0          0   
1         0           0                   0       0              0          0   
2         0           0                   0       0              0          0   
3         0           0                   0       0              0          0   
4         0           0                   0       0              0          0   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0         0                     0               0        0                0   
1         0                     0               0        0                0   
2         0                     0               0        0                0   
3         0                     0               0        0                0   
4         0                     0               0        0                0   

   Building  Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0         0             0              0         0        0     0   
1         0             0              0         0        0     0   
2         0             0              0         0        0     0   
3         0             0              0         0        0     0   
4    

In [166]:
toronto_onehot.shape

(1712, 237)

In [167]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Adult Boutique  \
0                            Adelaide, King, Richmond        0.000000   
1                                         Berczy Park        0.000000   
2        Brockton, Exhibition Place, Parkdale Village        0.000000   
3   Business Reply Mail Processing Centre 969 Eastern        0.000000   
4   CN Tower, Bathurst Quay, Island airport, Harbo...        0.000000   
5                         Cabbagetown, St. James Town        0.000000   
6                                  Central Bay Street        0.000000   
7           Chinatown, Grange Park, Kensington Market        0.000000   
8                                            Christie        0.000000   
9                                Church and Wellesley        0.011236   
10                     Commerce Court, Victoria Hotel        0.000000   
11                                         Davisville        0.000000   
12                                   Davisville North        0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...        0.000000   
14           Design Exchange, Toronto Dominion Centre        0.000000   
15                       Dovercourt Village, Dufferin        0.000000   
16             First Canadian Place, Underground city        0.000000   
17                Forest Hill North, Forest Hill West        0.000000   
18                     Harbord, University of Toronto        0.000000   
19  Harbourfront East, Toronto Islands, Union Station        0.000000   
20                          Harbourfront, Regent Park        0.000000   
21                      High Park, The Junction South        0.000000   
22                                      Lawrence Park        0.000000   
23                           Little Portugal, Trinity        0.000000   
24                        Moore Park, Summerhill East        0.000000   
25                                 North Toronto West        0.000000   
26                             Parkdale, Roncesvalles        0.000000   
27                                           Rosedale        0.000000   
28                                           Roselawn        0.000000   
29                                 Runnymede, Swansea        0.000000   
30                           Ryerson, Garden District        0.000000   
31                                     St. James Town        0.000000   
32                    Stn A PO Boxes 25 The Esplanade        0.000000   
33                                    Studio District        0.000000   
34                The Annex, North Midtown, Yorkville        0.000000   
35                                        The Beaches        0.000000   
36                     The Beaches West, India Bazaar        0.000000   
37                       The Danforth West, Riverdale        0.000000   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.000000            0.000000      0.000000   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.000000            0.000000      0.000000   
4            0.000000  0.071429            0.071429      0.071429   
5            0.000000  0.000000            0.000000      0.000000   
6            0.000000  0.000000            0.000000      0.000000   
7            0.000000  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.011236  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000            0.000000      0.000000   
15           0.000000  0.000000            0.000000      0.000000   
16           0.0000

In [168]:
toronto_grouped.shape

(38, 237)

In [169]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2           Steakhouse  0.04
3  American Restaurant  0.04
4      Thai Restaurant  0.04


----Berczy Park----
            venue  freq
0     Coffee Shop  0.07
1      Restaurant  0.05
2    Cocktail Bar  0.05
3          Bakery  0.04
4  Farmers Market  0.04


----Brockton, Exhibition Place, Parkdale Village----
                  venue  freq
0        Breakfast Spot  0.11
1                  Café  0.11
2           Coffee Shop  0.11
3          Climbing Gym  0.06
4  Caribbean Restaurant  0.06


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0         Yoga Studio  0.06
1  Light Rail Station  0.06
2          Comic Shop  0.06
3                Park  0.06
4          Restaurant  0.06


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0  Airport Terminal  

In [170]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [171]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighbourhood 1st Most Common Venue  \
0                            Adelaide, King, Richmond           Coffee Shop   
1                                         Berczy Park           Coffee Shop   
2        Brockton, Exhibition Place, Parkdale Village        Breakfast Spot   
3   Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4   CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   
5                         Cabbagetown, St. James Town           Coffee Shop   
6                                  Central Bay Street           Coffee Shop   
7           Chinatown, Grange Park, Kensington Market                  Café   
8                                            Christie                  Café   
9                                Church and Wellesley   Japanese Restaurant   
10                     Commerce Court, Victoria Hotel           Coffee Shop   
11                                         Davisville          Dessert Shop   
12                                   Davisville North          Burger Joint   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...           Coffee Shop   
14           Design Exchange, Toronto Dominion Centre           Coffee Shop   
15                       Dovercourt Village, Dufferin           Supermarket   
16             First Canadian Place, Underground city           Coffee Shop   
17                Forest Hill North, Forest Hill West                 Trail   
18                     Harbord, University of Toronto                  Café   
19  Harbourfront East, Toronto Islands, Union Station           Coffee Shop   
20                          Harbourfront, Regent Park           Coffee Shop   
21                      High Park, The Junction South    Mexican Restaurant   
22                                      Lawrence Park                  Park   
23                           Little Portugal, Trinity                   Bar   
24                        Moore Park, Summerhill East            Playground   
25                                 North Toronto West   Sporting Goods Shop   
26                             Parkdale, Roncesvalles        Breakfast Spot   
27                                           Rosedale                  Park   
28                                           Roselawn           Music Venue   
29                                 Runnymede, Swansea           Coffee Shop   
30                           Ryerson, Garden District           Coffee Shop   
31                                     St. James Town           Coffee Shop   
32                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
33                                    Studio District                  Café   
34                The Annex, North Midtown, Yorkville           Coffee Shop   
35                                        The Beaches           Coffee Shop   
36                     The Beaches West, India Bazaar        Sandwich Place   
37                       The Danforth West, Riverdale      Greek Restaurant   

   2nd Most Common Venue          3rd Most Common Venue 4th Most Common Venue  \
0                   Café                     Steakhouse   American Restaurant   
1           Cocktail Bar                     Restaurant    Seafood Restaurant   
2            Coffee Shop                           Café         Burrito Place   
3          Auto Workshop                           Park           Pizza Place   
4       Airport Terminal                Airport Service         Boat or Ferry   
5             Restaurant                    Pizza Place                  Café   
6                   Café             Italian Restaurant    Chinese Restaurant   
7                    Bar  Vegetarian / Vegan Restaurant                Bakery   
8          Grocery Store                           Park     Convenience Store   
9            Coffee Shop               Sushi Restaurant               Gay Bar   
10                  Café                     Restaurant            

In [172]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 2,
       0, 4, 0, 0, 1, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [173]:
neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village        Breakfast Spot   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse   American Restaurant   
1          Cocktail Bar            Restaurant    Seafood Restaurant   
2           Coffee Shop                  Café         Burrito Place   
3         Auto Workshop                  Park           Pizza Place   
4      Airport Terminal       Airport Service         Boat or Ferry   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Thai Restaurant        Clothing Store            Restaurant   
1            Steakhouse                Bakery    Italian Restaurant   
2               Stadium                   Bar  Caribbean Restaurant   
3            Restaurant               Butcher         Burrito Place   
4      Sculpture Garden       Harbor / Marina                 Plane   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   Hotel                   Bar                 Bakery  
1                Beer Bar                   Pub            Cheese Shop  
2  Furniture / Home Store          Climbing Gym     Italian Restaurant  
3                 Brewery            Skate Park             Smoke Shop  
4            Airport Gate    Airport Food Court                Airport

<h2>This code was tricky to figure out ... merging with a dataframe from way above that had not been used for many steps

In [174]:
toronto_merged = dftor

toronto_merged['Cluster Labels'] = kmeans.labels_

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Borough                   Neighbourhood   Latitude  Longitude  \
0     East Toronto                     The Beaches  43.676357 -79.293031   
1     East Toronto    The Danforth West, Riverdale  43.679557 -79.352188   
2     East Toronto  The Beaches West, India Bazaar  43.668999 -79.315572   
3     East Toronto                 Studio District  43.659526 -79.340923   
4  Central Toronto                   Lawrence Park  43.728020 -79.388790   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0           Coffee Shop                   Pub   
1               0      Greek Restaurant           Coffee Shop   
2               0        Sandwich Place                  Park   
3               0                  Café           Coffee Shop   
4               0                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Music Venue          Neighborhood           Yoga Studio   
1        Ice Cream Shop             Bookstore    Italian Restaurant   
2            Steakhouse      Sushi Restaurant         Movie Theater   
3    Italian Restaurant                Bakery   American Restaurant   
4    Dim Sum Restaurant              Bus Line           Yoga Studio   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Dog Run  Fast Food Restaurant        Farmers Market   
1           Yoga Studio        Cosmetics Shop               Brewery   
2               Brewery                   Pub          Intersection   
3           Yoga Studio                  Park    Seafood Restaurant   
4      Doner Restaurant   Filipino Restaurant  Fast Food Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0    Falafel Restaurant            Event Space  
1       Bubble Tea Shop             Restaurant  
2     Food & Drink Shop         Ice Cream Shop  
3        Sandwich Place            Cheese Shop  
4        Farmers Market     Falafel Restaurant

In [186]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [176]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
0                                         The Beaches           Coffee Shop   
1                        The Danforth West, Riverdale      Greek Restaurant   
2                      The Beaches West, India Bazaar        Sandwich Place   
3                                     Studio District                  Café   
4                                       Lawrence Park                  Park   
5                                    Davisville North          Burger Joint   
6                                  North Toronto West   Sporting Goods Shop   
7                                          Davisville          Dessert Shop   
8                         Moore Park, Summerhill East            Playground   
9   Deer Park, Forest Hill SE, Rathnelly, South Hi...           Coffee Shop   
10                                           Rosedale                  Park   
11                        Cabbagetown, St. James Town           Coffee Shop   
12                               Church and Wellesley   Japanese Restaurant   
13                          Harbourfront, Regent Park           Coffee Shop   
14                           Ryerson, Garden District           Coffee Shop   
15                                     St. James Town           Coffee Shop   
16                                        Berczy Park           Coffee Shop   
18                           Adelaide, King, Richmond           Coffee Shop   
19  Harbourfront East, Toronto Islands, Union Station           Coffee Shop   
20           Design Exchange, Toronto Dominion Centre           Coffee Shop   
21                     Commerce Court, Victoria Hotel           Coffee Shop   
23                Forest Hill North, Forest Hill West                 Trail   
25                     Harbord, University of Toronto                  Café   
26          Chinatown, Grange Park, Kensington Market                  Café   
29             First Canadian Place, Underground city           Coffee Shop   
30                                           Christie                  Café   
31                       Dovercourt Village, Dufferin           Supermarket   
32                           Little Portugal, Trinity                   Bar   
33       Brockton, Exhibition Place, Parkdale Village        Breakfast Spot   
34                      High Park, The Junction South    Mexican Restaurant   
35                             Parkdale, Roncesvalles        Breakfast Spot   
36                                 Runnymede, Swansea           Coffee Shop   
37  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   

   2nd Most Common Venue          3rd Most Common Venue 4th Most Common Venue  \
0                    Pub                    Music Venue          Neighborhood   
1            Coffee Shop                 Ice Cream Shop             Bookstore   
2                   Park                     Steakhouse      Sushi Restaurant   
3            Coffee Shop             Italian Restaurant                Bakery   
4            Swim School             Dim Sum Restaurant              Bus Line   
5         Sandwich Place                 Breakfast Spot                   Gym   
6            Coffee Shop                    Yoga Studio        Sandwich Place   
7         Sandwich Place                    Pizza Place           Coffee Shop   
8                    Gym                          Trail        Discount Store   
9                    Pub               Sushi Restaurant     Convenience Store   
10            Playground                          Trail                 Diner   
11            Restaurant                    Pizza Place                  Café   
12           Coffee Shop               Sushi Restaurant               Gay Bar   
13                   Pub                           Park                Bakery   
14        Clothing Store                           Café        Cosmetics Shop   
15            Restaurant                           Café  

In [177]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
17                                 Central Bay Street           Coffee Shop   
27  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17                  Café    Italian Restaurant    Chinese Restaurant   
27      Airport Terminal       Airport Service         Boat or Ferry   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17                   Bar          Burger Joint     Indian Restaurant   
27      Sculpture Garden       Harbor / Marina                 Plane   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17    Falafel Restaurant       Bubble Tea Shop        Thai Restaurant  
27          Airport Gate    Airport Food Court                Airport

In [178]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
22      Roselawn           Music Venue                Garden   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
22           Yoga Studio           Fish Market   Filipino Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
22  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

   9th Most Common Venue 10th Most Common Venue  
22           Event Space   Ethiopian Restaurant

In [179]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
28  Stn A PO Boxes 25 The Esplanade           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
28            Restaurant                  Café    Italian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
28          Cocktail Bar    Seafood Restaurant                   Pub   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
28                 Hotel              Beer Bar               Creperie

In [180]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
24  The Annex, North Midtown, Yorkville           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
24                  Café        Sandwich Place           Pizza Place   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
24                  Park             BBQ Joint         Grocery Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
24     Indian Restaurant     Jewish Restaurant            Flower Shop